Algorithm.Framework/QCAlgorithmFramework.cs

```diff
@@ -0,0 +1,202 @@
+using QuantConnect.Data;
+using QuantConnect.Data.UniverseSelection;
+
+namespace QuantConnect.Algorithm.Framework
+{
+    /// <summary>
+    /// Algorithm framework base class that enforces a modular approach to algorithm development
+    /// </summary>
+    public class QCAlgorithmFramework : QCAlgorithm
+    {
+        /// <summary>
+        /// Gets or sets the portfolio selection model.
+        /// </summary>
+        public IPortfolioSelectionModel PortfolioSelection { get; set; }
+
+        /// <summary>
+        /// Gets or sets the signal model
+        /// </summary>
+        public ISignalModel Signal { get; set; }
+
+        /// <summary>
+        /// Gets or sets the portoflio construction model
+        /// </summary>
+        public IPortfolioConstructionModel PortfolioConstruction { get; set; }
+
+        /// <summary>
+        /// Gets or sets the execution model
+        /// </summary>
+        public IExecutionModel Execution { get; set; }
+
+        /// <summary>
+        /// Gets or sets the risk management model
+        /// </summary>
+        public IRiskManagementModel RiskManagement { get; set; }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="QCAlgorithmFramework"/> class
+        /// </summary>
+        public QCAlgorithmFramework()
+        {
+            var type = GetType();
+            var frameworkOnData = type.GetMethod(nameof(OnFrameworkData), new[] {typeof(Slice)});
+            if (frameworkOnData.DeclaringType != typeof(QCAlgorithmFramework))
+            {
+                throw new Exception("Framework algorithms can not override OnFrameworkData(Slice)");
+            }
+            var frameworkOnSecuritiesChanged = type.GetMethod(nameof(OnFrameworkSecuritiesChanged), new[] {typeof(SecurityChanges)});
+            if (frameworkOnSecuritiesChanged.DeclaringType != typeof(QCAlgorithmFramework))
+            {
+                throw new Exception("Framework algorithms can not override OnFrameworkSecuritiesChanged(SecurityChanges)");
+            }
+
+            // set model defaults
+            Execution = new ImmediateExecutionModel();
+            RiskManagement = new NullRiskManagementModel();
+        }
+
+        /// <summary>
+        /// Called by setup handlers after Initialize and allows the algorithm a chance to organize
+        /// the data gather in the Initialize method
+        /// </summary>
+        public override void PostInitialize()
+        {
+            CheckModels();
+
+            foreach (var universe in PortfolioSelection.CreateUniverses(this))
+            {
+                AddUniverse(universe);
+            }
+
+            base.PostInitialize();
+        }
+
+        /// <summary>
+        /// Used to send data updates to algorithm framework models
+        /// </summary>
+        /// <param name="slice">The current data slice</param>
+        public override void OnFrameworkData(Slice slice)
```

Comment:
`OnFrameworkData` and `OnFrameworkSecuritiesChanged` can be declared `sealed` to prevent overriding them in user algorithms. This could avoid using reflection in the constructor to check if the methods have been overridden.

Reply:
That's a great idea and much better!